gapic/templates/%namespace/%name_%version/%sub/services/%service/_shared_macros.j2

```diff
@@ -55,6 +55,47 @@ except ImportError:  # pragma: NO COVER
 {% endif %}{# service_version #}
 {% endmacro %}
 
+{% macro create_metadata(method) %}
+    {% if method.explicit_routing %}
+        header_params = {}
+        {% if not method.client_streaming %}
+        {% for routing_param in method.routing_rule.routing_parameters %}
+        {% if routing_param.path_template %} {# Need to match. #}
+
+        routing_param_regex = {{ routing_param.to_regex() }}
+        regex_match = routing_param_regex.match(request.{{ routing_param.field }})
+        if regex_match and regex_match.group("{{ routing_param.key }}"):
+            header_params["{{ routing_param.key }}"] = regex_match.group("{{ routing_param.key }}")
+
+    {% else %}
+
+        if request.{{ routing_param.field }}:
+            header_params["{{ routing_param.key }}"] = request.{{ routing_param.field }}
+
+        {% endif %}
+        {% endfor %} {# method.routing_rule.routing_parameters #}
+        {% endif %} {# if not method.client_streaming #}
+
+        if header_params:
+            metadata = tuple(metadata) + (
+                gapic_v1.routing_header.to_grpc_metadata(header_params),
+            )
+
+    {% elif method.field_headers %}{# implicit routing #}
+        # Certain fields should be provided within the metadata header;
+        # add these here.
+        metadata = tuple(metadata) + (
+            gapic_v1.routing_header.to_grpc_metadata((
+                {% for field_header in method.field_headers %}
+                {% if not method.client_streaming %}
```

Comment:
nit: I know this PR just moved the pre-existing code into this macro, but we could move the `{% if %}`  outside the `{% for %}` since `method` is not altered by the loop.

Reply:
Fixed in https://github.com/googleapis/gapic-generator-python/pull/2133/commits/3da4cc9436b84863dfa917ba3061aff05185ceb6